akka-http-core/src/main/scala/akka/http/impl/engine/rendering/HttpResponseRendererFactory.scala

```diff
@@ -33,22 +33,22 @@ private[http] class HttpResponseRendererFactory(serverHeader: Option[headers.Ser
     }
 
   // as an optimization we cache the Date header of the last second here
-  @volatile private[this] var cachedDateHeader: (Long, Array[Byte]) = (0L, null)
-
+  private[this] var cachedDateHeader: (Long, Array[Byte]) = (0L, null)
   private def dateHeader: Array[Byte] = {
-    var (cachedSeconds, cachedBytes) = cachedDateHeader
-    val now = System.currentTimeMillis
-    if (now - 1000 > cachedSeconds) {
-      cachedSeconds = now / 1000
+    var cachedSeconds = cachedDateHeader._1
+    val now = currentTimeMillis()
+    val nowSeconds = now / 1000
+    if (nowSeconds > cachedSeconds) {
+      cachedSeconds = nowSeconds
       val r = new ByteArrayRendering(48)
-      dateTime(now).renderRfc1123DateTimeString(r ~~ headers.Date) ~~ CrLf
-      cachedBytes = r.get
-      cachedDateHeader = cachedSeconds -> cachedBytes
+      DateTime(now).renderRfc1123DateTimeString(r ~~ headers.Date) ~~ CrLf
+      cachedDateHeader = cachedSeconds -> r.get
     }
-    cachedBytes
+    cachedDateHeader._2
```

Comment:
If you switch to reading the `cachedSeconds` and `cachedBytes` independently then we could also switch to writing them independently and get rid of the tuple altogether, no?